packages/drivers/routerlicious-driver/src/sessionInfoManager.ts

```
@@ -0,0 +1,145 @@
+	/**
+	 * Stored session info
+	 * Key: URL for given session (see "getDiscoverSessionUrl")
+	 * Value: session info stored as an IResolvedUrl
+	 */
+	private readonly sessionFluidUrls: Map<string, IResolvedUrl> = new Map();
+
+	/**
+	 * Stored dates of when a session was last discovered/refreshed
+	 * Key: URL for given session (see "getDiscoverSessionUrl")
+	 * Value: date last discovered
+	 */
+	private readonly sessionLastDiscovered: Map<string, number> = new Map();
+
+	constructor(private readonly enableDiscovery: boolean) {}
+
+	/**
+	 * Start tracking info for a given session
+	 */
+	public async initializeSessionInfo(
+		params: IGetSessionInfoParams & { session: ISession | undefined },
+	): Promise<IResolvedUrl> {
+		const { resolvedUrl, session } = params;
+
+		const url = getDiscoverSessionUrl(params);
+		assert(
+			this.sessionFluidUrls.has(url) === this.sessionLastDiscovered.has(url),
+			"Session map state mismatch",
+		);
+
+		if (session !== undefined) {
+			this.sessionFluidUrls.set(url, getDiscoveredFluidResolvedUrl(resolvedUrl, session));
+			this.sessionLastDiscovered.set(url, Date.now());
+		} else if (!this.sessionFluidUrls.has(url)) {
+			this.sessionFluidUrls.set(url, resolvedUrl);
+			// Force a refresh
+			this.sessionLastDiscovered.set(url, 0);
+		}
+
+		return (await this.getSessionInfo(params)).resolvedUrl;
+	}
+
+	/**
+	 * Retrieve, and potentially refresh, info of a given session
+	 */
+	public async getSessionInfo(
+		params: IGetSessionInfoParams,
+	): Promise<IGetSessionInfoResponse> {
+		const url = getDiscoverSessionUrl(params);
+		assert(
+			this.sessionFluidUrls.has(url) && this.sessionLastDiscovered.has(url),
+			"Unexpected discover session URL",
+		);
+
+		let refreshed = false;
+		const shouldRediscover =
+			Date.now() - this.sessionLastDiscovered.get(url)! > RediscoverAfterTimeSinceDiscoveryMs;
+		if (this.enableDiscovery && shouldRediscover) {
+			await this.updateSessionInfo(params).catch(() => {
```

Posing this question to no one specific:
In the original code, this `catch` clause didn't rethrow the error so I kept that behavior. Do we care this isn't re-thrown and we hide the error?

---

For visibility, hiding the error caused some tests to fail: https://github.com/microsoft/FluidFramework/pull/22330.

Fix: https://github.com/microsoft/FluidFramework/pull/22335